share-coordinator/src/main/java/org/apache/kafka/coordinator/share/StateBatchUtil.java

```diff
@@ -0,0 +1,360 @@
+                            // candidate has higher priority
+                            sortedBatches.add(new PersisterStateBatch(
+                                last.firstOffset(),
+                                candidate.firstOffset() - 1,
+                                last.deliveryState(),
+                                last.deliveryCount()
+                            ));
+
+                            sortedBatches.add(candidate);
+                        }
+                    }
+                }
+            }
+            overlapState = getOverlappingState(sortedBatches);
+        }
+        finalBatches.addAll(sortedBatches);   // some non overlapping batches might have remained
+        return finalBatches;
+    }
+
+    /**
+     * Accepts a sorted set of state batches and finds the first 2 batches which overlap.
+     * Overlap means that they have some offsets in common or, they are contiguous with the same state.
+     * Along with the 2 overlapping batches, also returns a list of non overlapping intervals
+     * prefixing them. For example
+     * _____ ____  _____ _____      _____
+     *                      ______     __
+     * <---------------> <-------->
+     *  non-overlapping   1st overlapping pair
+     *
+     * @param sortedBatches - TreeSet representing sorted set of {@link PersisterStateBatch}
+     * @return object of {@link BatchOverlapState} representing overlapping pair and non-overlapping prefix
+     */
+    private static BatchOverlapState getOverlappingState(TreeSet<PersisterStateBatch> sortedBatches) {
+        if (sortedBatches == null || sortedBatches.isEmpty()) {
+            return BatchOverlapState.SENTINEL;
+        }
+        Iterator<PersisterStateBatch> iter = sortedBatches.iterator();
+        PersisterStateBatch last = iter.next();
+        List<PersisterStateBatch> nonOverlapping = new ArrayList<>(sortedBatches.size());
+        while (iter.hasNext()) {
+            PersisterStateBatch candidate = iter.next();
+            if (candidate.firstOffset() <= last.lastOffset() || // overlap
+                last.lastOffset() + 1 == candidate.firstOffset() && compareBatchState(last, candidate) == 0) {  // contiguous
+                return new BatchOverlapState(
+                    last,
+                    candidate,
+                    nonOverlapping
+                );
+            }
+            nonOverlapping.add(last);
+            last = candidate;
+        }
+        return new BatchOverlapState(null, null, nonOverlapping);
+    }
+
+    /**
+     * Compares the state of 2 batches i.e. the deliveryCount and deliverState.
+     * Uses standard compareTo contract x < y => +int, x > y => -int, x == y => 0
+     * @param b1 - {@link PersisterStateBatch} to compare
+     * @param b2 - {@link PersisterStateBatch} to compare
+     * @return int representing comparison result.
+     */
+    private static int compareBatchState(PersisterStateBatch b1, PersisterStateBatch b2) {
```

Comment:
We use the offsets to determine the relative positioning and then use just that state to determine how to break the batches. Based on state, overlapping batches could be merged or broken down.
example:
1. 
------               [1,10,0,1]
    -------          [5,15,0,1]
=>
----------          [1,15,0,1]

2.
------               [1,10,0,1]
    -------          [5,15,2,1]
=>
----          [1,4,0,1]
     ------- [5,15,2,1]

Reply:
The idea behind delivery count taking precedence is because it has the connotation of epoch. Higher number => higher chance of it being fresher.
If 2 batches have same delivery count then some states which are considered terminal like ACK or ARCHIVED take precedence. Since the ordinal in the delivery state enum is numerically higher for these, they blend in well.
Betweek ACK and ARCHIVED, it is acceptable to take any (we take ARCHIVED).

HOWEVER, ARCHIVED state in not implemented in kip-932

Reply:
I think ARCHIVING is not implemented in KIP_932, and that's because it's going to be in the DLQ KIP to follow.